anvil/src/buffer_utils.rs

```diff
@@ -0,0 +1,59 @@
+use std::{cell::RefCell, rc::Rc};
+
+use slog::Logger;
+
+#[cfg(feature = "egl")]
+use smithay::backend::egl::{BufferAccessError, EGLDisplay};
+use smithay::{
+    reexports::wayland_server::protocol::wl_buffer::WlBuffer,
+    wayland::shm::with_buffer_contents as shm_buffer_contents,
+};
+
+/// Utilities for working with `WlBuffer`s.
+#[derive(Clone)]
+pub struct BufferUtils {
+    #[cfg(feature = "egl")]
+    egl_display: Rc<RefCell<Option<EGLDisplay>>>,
+    log: Logger,
+}
+
+impl BufferUtils {
+    /// Creates a new `BufferUtils`.
+    #[cfg(feature = "egl")]
+    pub fn new(egl_display: Rc<RefCell<Option<EGLDisplay>>>, log: Logger) -> Self {
+        Self { egl_display, log }
+    }
+
+    /// Creates a new `BufferUtils`.
+    #[cfg(not(feature = "egl"))]
+    pub fn new(log: Logger) -> Self {
+        Self { log }
+    }
+
+    /// Returns the dimensions of an image stored in the buffer.
+    #[cfg(feature = "egl")]
+    pub fn dimensions(&self, buffer: WlBuffer) -> Option<(i32, i32)> {
```

Comment:
I think this method (and `shm_buffer_dimensions`) can be changed to only require a `&WlBuffer` as well, and avoid unneeded clones.